packages/driver-adapter-utils/src/types.ts

```diff
@@ -129,70 +129,120 @@ const officialPrismaAdapters = [
   '@prisma/adapter-pg-worker',
 ] as const
 
-export interface Queryable {
-  readonly provider: Provider
-  readonly adapterName: (typeof officialPrismaAdapters)[number] | (string & {})
+/**
+ * A generic driver adapter that allows the user to connect to a
+ * database. The query and result types are specific to the adapter.
+ */
+export interface DriverAdapter<Query, Result> extends AdapterInfo {
+  /**
+   * Connect to the database.
+   */
+  connect(): Promise<Queryable<Query, Result>>
+}
 
+export interface SqlDriverAdapter extends DriverAdapter<SqlQuery, SqlResultSet> {
+  connect(): Promise<SqlConnection>
+}
+
+/**
+ * An SQL migration adapter that is aware of the notion of a shadow database
+ * and can create a connection to it.
+ */
+export interface SqlMigrationAwareDriverAdapter extends SqlDriverAdapter {
+  connectToShadowDb(): Promise<SqlConnection>
+}
+
+export interface Queryable<Query, Result> extends AdapterInfo {
   /**
-   * Execute a query given as SQL, interpolating the given parameters,
-   * and returning the type-aware result set of the query.
-   *
-   * This is the preferred way of executing `SELECT` queries.
+   * Execute a query and return its result.
    */
-  queryRaw(params: Query): Promise<Result<ResultSet>>
+  queryRaw(params: Query): Promise<Result>
 
   /**
-   * Execute a query given as SQL, interpolating the given parameters,
-   * and returning the number of affected rows.
-   *
-   * This is the preferred way of executing `INSERT`, `UPDATE`, `DELETE` queries,
-   * as well as transactional queries.
+   * Execute a query and return the number of affected rows.
    */
-  executeRaw(params: Query): Promise<Result<number>>
+  executeRaw(params: Query): Promise<number>
 }
 
-export interface TransactionContext extends Queryable {
+export interface SqlQueryable extends Queryable<SqlQuery, SqlResultSet> {}
+
+export interface SqlConnection extends SqlQueryable {
   /**
-   * Starts new transaction.
+   * Execute multiple SQL statements separated by semicolon.
    */
-  startTransaction(): Promise<Result<Transaction>>
-}
+  executeScript(script: string): Promise<void>
 
-export interface DriverAdapter extends Queryable {
   /**
-   * Starts new transaction.
+   * Start new transaction.
    */
-  transactionContext(): Promise<Result<TransactionContext>>
+  transactionContext(): Promise<TransactionContext>
 
   /**
    * Optional method that returns extra connection info
    */
-  getConnectionInfo?(): Result<ConnectionInfo>
+  getConnectionInfo?(): ConnectionInfo
+
+  /**
+   * Dispose of the connection and release any resources.
+   */
+  dispose(): Promise<void>
+}
+
+export interface TransactionContext extends AdapterInfo, SqlQueryable {
+  /**
+   * Starts new transaction.
```

Comment:
```suggestion
   * Start new transaction.
```

To make it consistent with other changes elsewhere.